### PR TITLE
Adds tests

### PR DESCRIPTION
**Pull request by dm36**
Improve coveralls coverage by testing updates and error-handling.

Ref: LOG-7721
Semver: Patch

### CODE REVIEW

File: logdna/resource_view_test.go
Comment (nimbinatus) [inline logdna/resource_view_test.go:184]: So in these tests, I don't see any expected failure tests&mdash;explicit failure cases such as passing a boolean instead of a string or an invalid value instead of a valid one for values that are constrained to a set. I'm not comfortable approving a PR without those, personally, as those failure cases need to be handled gracefully to ensure a strong developer experience.

**Review by nimbinatus — CHANGES_REQUESTED** [review]

File: logdna/resource_view_test.go
Comment (dm36) [reply logdna/resource_view_test.go:184]: @nimbinatus all the expected failure tests can be found here: https://github.com/logdna/terraform-provider-logdna/blob/8f89c85fe571dd54e432558ad8edbe4440e9a031/logdna/resource_view_test.go#L12

Comment (nimbinatus) [reply logdna/resource_view_test.go:184]: I see. I don't see a failure test for a view, however. Maybe this isn't a thing in Go, but I suspect it is. I'm expecting something like this, but in actual Go code and not my poor pseudocode:
```go
func TestAssertErrorViewBulkEmails(t *testing.T err error) {
	name := "test"
	query := "test"
	category := "DEMO"
	tags1 := "host1"
	tags2 := 123

	resource.Test(t, resource.TestCase{
		Providers: testAccProviders,
		Steps: []resource.TestStep{
			{
				Config: testViewConfigBulkEmails(name, query, category, tags1, tags2),
				Check: resource.ComposeTestCheckFunc(
					testViewExists("logdna_view.new"),
					resource.TestCheckResourceAttr("logdna_view.new", "name", name),
					resource.TestCheckResourceAttr("logdna_view.new", "query", query),
					resource.TestCheckResourceAttr("logdna_view.new", "category.#", "1"),
					resource.TestCheckResourceAttr("logdna_view.new", "category.0", category),
					resource.TestCheckResourceAttr("logdna_view.new", "tags.#", "2"),
					resource.TestCheckResourceAttr("logdna_view.new", "tags.0", tags1),
					resource.TestCheckResourceAttr("logdna_view.new", "tags.1", tags2),
					resource.TestCheckResourceAttr("logdna_view.new", "webhook_channels.#", "0"),
				),
			},
		},
	})
        if assert.Error(t, err) {
	      assert.Equal(t, expectedError, err)
        }
        else {
            t.Fail()
	    t.Logf("View created with invalid values %s", tags2) # Though I'd also expect the single string for `categories` to fail as it's not an array.
        }
}
```

Personally, I'd also add tests to validate each string for the key-value pairs that have set values as options. For example, what if someone put `pagerduty_channel.0.immediate` as `"no"` instead of `"false"`&mdash;or if they mistyped and entered `"fasle"`, which I've done myself many times? How do you test for a graceful failure or error here? We'd rather see that kind of mistake just error out with a reasonable message for the user versus possibly crashing or giving the user no feedback at all.

Comment (dm36) [reply logdna/resource_view_test.go:184]: @nimbinatus am following the guidelines provided here on how to test error handling with the terraform sdk: https://www.terraform.io/docs/extend/best-practices/testing.html#expecting-errors-or-non-empty-plans:

You can see in the test here is the user doesn't provide a servicekey the provider will give the following error:

`
ExpectError: regexp.MustCompile("The argument \"servicekey\" is required, but no definition was found."),
`

Comment (dm36) [reply logdna/resource_view_test.go:184]: also for the situation as described you'll get the following error: 

Screen Shot 2020-10-26 at 6 44 22 PM

Am not able to assert on the expected on an error of that sort with the terraform sdk.

**Comment by mdeltito** [comment]
this is looking good to me - I'm not finding any info on testing http-related errors via the builtin `Test` helpers, but tests for `client.go` could be written in a terraform-agnostic way (e.g. using `httptest` server).

File: logdna/resource_view_test.go
Comment (mdeltito) [inline logdna/resource_view_test.go:309]: This pattern of adding infinite function parameters to a templating seems pretty unmaintainable.  The test config is very static, even for the fields that are using the values passed in - there are always 2 apps, 2 hosts, etc. So the added indirection of a function wrapper around `fmt.Sprintf` doesn't really do much for us, aside from providing a place to store the template string. I don't think we're gaining much here by having this as a function vs just defining the template string as a top-level variable and calling `fmt.Sprintf` directly.

I don't think it's necessary to change in this PR, just a pattern I think we should find an alternative for in future work.

File: logdna/resource_view_test.go
Comment (mdeltito) [reply logdna/resource_view_test.go:184]: @dm36 RE that screenshot, it looks like you'd just need to adjust the regex to match on the error you're looking for? It looks like it should work to use `ExpectError` in that way and testing that the user gets the expected error output - the extra debug log output in the error text might be throwing you off. Can that debug output be disabled under test?

you also likely need to escape (at minimum) the bracket characters in the `regexp.MustCompile`
`Error: "channels[0].immediate" must be...`

**Review by mdeltito — APPROVED** [review]
A couple of comments, I think this is looking good though. I still think we could use some tests around the error cases in `client.go`

File: logdna/resource_view_test.go
Comment (dm36) [reply logdna/resource_view_test.go:184]: Yeah I'm not sure how to make assertions on errors returned from the config-api with the terraform sdk and the ExpectError function. I've also created a post here: https://discuss.hashicorp.com/t/testing-errors-when-building-a-terraform-provider/16770

Comment (dm36) [reply logdna/resource_view_test.go:184]: can configure terraform error levels but even the lowest level still has that debug message at the beginning: https://www.terraform.io/docs/internals/debugging.html

Screen Shot 2020-10-27 at 11 34 31 PM

File: logdna/resource_view_test.go
Comment (dm36) [reply logdna/resource_view_test.go:309]: how should we try to change this in the future? passing along the values in the sprintf also makes sense in the context that I can assert that the values match the variables in the TestCheckResourcesAttr but I also do agree that the # of variables is a lot for a couple of the tests

Comment (dm36) [reply logdna/resource_view_test.go:309]: i could not sprintf and pass anything to the config files but still have the variables as reference and assert on them in the checkresourcesattr

Comment (mdeltito) [reply logdna/resource_view_test.go:309]: yeah you could have the template string as a top-level variable and use `fmt.Sprintf` where you need it. just removing a layer:


```golang
var testConfigBulkEmails = `provider "logdna" {
	}
	resource "logdna_view" "new" {
		name = "%s"
		query = "%s"
	}`

func TestViewBulkEmailsUpdate(t *testing.T) {
    foo := "foo"
    bar := "bar"
    
    config := fmt.Sprintf(testConfigBulkEmails, foo, bar)
    ...
```
there might be other ways to manage all of this too, I'm just thinking of the simplest form.

File: logdna/resource_view_test.go
Comment (nimbinatus) [reply logdna/resource_view_test.go:184]: After talking with some folks at Hashicorp, the `ExpectError` is the right thing to use. For the most part, people test for the happy case, but it isn't necessarily uncommon to test for errors as I noted. They sent along samples like https://github.com/terraform-providers/terraform-provider-aws/search?q=expecterror and https://github.com/hashicorp/terraform-provider-google/search?q=expecterror.

I think that, in this case, the `ExpectError` is the assert, so it is the way to go. Some similar ideas to look at:
- https://github.com/hashicorp/terraform-provider-google/blob/b0e3ae2a9e7235af180fc1c06380659d0b3a474f/google/resource_bigtable_instance_test.go#L27
- A bit more involved, using the `TestCheckFunc` ([docs](https://godoc.org/github.com/hashicorp/terraform-plugin-sdk/helper/resource#TestCheckFunc)) [here](https://github.com/terraform-providers/terraform-provider-aws/blob/a62eec8d6752388cebbb62595bc037318de23286/aws/resource_aws_ses_domain_identity_verification_test.go#L81-L128) called [here](https://github.com/terraform-providers/terraform-provider-aws/blob/a62eec8d6752388cebbb62595bc037318de23286/aws/resource_aws_ses_domain_identity_verification_test.go#L39)

Comment (nimbinatus) [reply logdna/resource_view_test.go:184]: I'm also okay with dropping this for a later PR if @mdeltito is good with that.

Comment (mdeltito) [reply logdna/resource_view_test.go:184]: yeah I think this is good - we have test coverage for error cases on [custom validators](https://github.com/logdna/terraform-provider-logdna/blob/9035ef495c2405f07b99115a3905eb6dba5571b5/logdna/resource_view_test.go#L94-L116) (among other things) which I think is the most important. 

@dm36 we'll need to sync up on what the issue was with using `ExpectError` for other error cases at a later point, I'm not sure I really understand based on your screenshots after testing these things locally.

**Review by mdeltito — APPROVED** [review]

**Review by nimbinatus — COMMENTED** [review]
I'm good with it as long as Mike is, but I'd wait for further approvals from the other engineers before merging.

**Review by ydshah2 — APPROVED** [review]
looks good. LOG-7779 made for followup task error handling support.

**Pull request merged**
